### DISCUSSION TITLE
How do I limit game to the "safe area" on iOS?

### DISCUSSION THREAD

**Discussion by divVerent** [hajimehoshi/ebiten#3143] [Q&A] [answered]
I am making an application in Ebiten UI (Ebiten chosen mainly because it runs on HTML5 including desktop and Android _and_ on iOS).

It's already kinda working, except that part of the UI overlaps with camera cutouts and UI controls:

<img width="480" alt="Screenshot 2024-10-21 at 1 04 02 PM" src="https://github.com/user-attachments/assets/c869e133-5a85-4b02-9f75-8b4afe396e1b">

I discovered this concept of a "safe area" in Xcode:

<img width="1728" alt="Screenshot 2024-10-21 at 1 05 18 PM" src="https://github.com/user-attachments/assets/310e54ac-f9a0-41d4-a61d-1937d9e36204">

However, I can't seem to find any way to constrain the MTKView that the GameViewController owns to this safe area. The storyboard editor _looks_ like it should be able to do that, but I can't figure out how - the very promising "Add New Constraints" dialog I can open at the bottom right is all greyed out as well:

<img width="315" alt="Screenshot 2024-10-21 at 1 06 50 PM" src="https://github.com/user-attachments/assets/0af3d0f7-5c62-4ae6-83b3-5edd91eb8756">

How does one do that?

(Sadly I can't run the HTML5 version on iOS - I'd actually prefer that, but Safari doesn't support the Web MIDI API...)

**Comment by hajimehoshi** [comment, marked as answer]
What about getting the safe are from the Objective-C world and inject it to the Ebitengine world?
https://stackoverflow.com/questions/47075615/objective-c-how-to-create-self-view-inside-safe-area-programmatically

**Reply by divVerent** [reply]
I guess that would work, but is that really the intended way? The Xcode editor there totally _feels_ like I can somehow make a view whose bounds are limited by the safe area, and stick all of Ebitengine into that view - I just don't seem to understand how (like, can't even figure out how to place _anything_ in the container view object that totally _exists_).

Like, I can make a container view, and there's a system of constraints, and from what I read, it can honor the safe area. So all I need to know is how to do that - or failing that, maybe it's easier to just write the XML structure of what I need?

Going this route might work, but I will have to see what it takes to make Ebiten UI restrict itself to the smaller size (e.g. by using AnchorLayout and reconfiguring it when the safe area changes: https://github.com/ebitenui/ebitenui/blob/master/_examples/widget_demos/anchorlayout/main.go#L32).

**Reply by divVerent** [reply]
Actually, instead of dealing with AnchorLayout mess, reading Ebiten UI source, maybe I can just pass a SubImage in Draw() and it's happy: https://github.com/ebitenui/ebitenui/blob/master/ui.go#L79 appears to be what is propagating the screen size down to the widgets.

Edit: SubImage approach doesn't work, as Ebiten UI can only deal with images that have the origin at 0, 0. Otherwise the entire UI is shifted.

**Reply by divVerent** [reply]
Thanks - this appears to work indeed.

Seems one can't reconfigure AnchorLayout once created, but I can just recreate the entire UI when insets change (already doing that on resolution change anyway). As both events are primarily triggered by screen rotation and tend to happen in the same frame, that's fine.

I still think it's some nasty hackery to do it this way and there must be a way to do it with the storyboard... but meh. This works. Thanks!
